HW5

avramirs opened the pull request:
Пока реализовано оторвано от карточки, в рамках этого дз хочу прикрутить скиллс к ней, а также залить это решение на TS, если успею, считаю, что сейчас надо упростить js, чтобы переписать потом на ts. Хочу также попросить прощения за столь позднее открытие PR'а, но надеюсь, что ещё не поздно.
Страница подключена к gh-pages: https://ok-technopolis-2021.github.io/Kovalenko_Pavel/

avramirs committed:
Initialize hw5

avramirs committed:
Replace index.html due to gh-pages' requires

avramirs committed:
Remove duplicate of index.html

PaaavelZ commented:
Страница подключена к gh-pages: https://ok-technopolis-2021.github.io/Kovalenko_Pavel/

avramirs committed:
Fix problem with displaying rubbish bin near skill progress range

avramirs commented:
Изначально js делал в двух файлах - skill.js и skillStorage.js, однако возникла проблема с тем, что skillStorage нужен был класс skill, но заимпортить я его никак не смог - пытался и просто import "...", и @import("..."), однако ничего не помогало. В итоге разбил один класс - skillStorage - с концами, ибо два класса в файле не айс

avramirs commented:
Серия файлов sliders отвечает исключительно за отрисовку слайдеров, выделил их особняком, ибо брал с интернета уже сгенерированный под мои нужды код, за весь который не ручаюсь и не могу в полной мере осознать, есть ли там ненужные строчки и тд тп.

avramirs commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+        let skillRatio = Number.parseInt(Skill.#skillRatioInput.value);
+        let minSkillRatio = Number.parseInt(Skill.#skillRatioInput.min);
+        let maxSkillRatio = Number.parseInt(Skill.#skillRatioInput.max);
+        if (isNaN(skillRatio) || skillRatio < minSkillRatio || skillRatio > maxSkillRatio) {
```
Вот тут, наверно, в кейсах, когда введённый уровень скила не находится в диапазоне между min и max, надо его насильно делать либо min, либо max, а не кидать undefined.

avramirs replied:
✅

avramirs commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+
+    static #renderDeleteSkillBtn(skill) {
+        const deleteSkillButton = document.createElement("input");
+        deleteSkillButton.type = "image";
```
Пытался сделать html-кодом напрямую, как в методе getSkillHeaderInnerHtml немного выше этого, однако там отличие в том, что мы описываем inner html уже созданного элемента, а тут, по сути, описываем html уже готового элемента, инпута, из-за чего DOMParser как-то не так отрабатывает и всё летит в тар-тарары

BusinessDuck replied:
Переносы строк появляюстя пустые если делать готовым HTML ?

avramirs replied:
Видимо, нервы не сдержали. Нашёл один рабочий способ через дом парсер, уже реализовываю

avramirs replied:
✅

avramirs commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+    isPlugRendered = false;
+}
+
+document.getElementById("skills__header__add-skill-wrap")
```
Вот тут накостылял, не знал об input type="image", подправлю на него, и там уже можно будет по onclick действия описывать

avramirs replied:
✅

avramirs commented on index.html:
```diff
@@ -0,0 +1,30 @@
+        <img class="skills__header__add-skill" id="skills__header__add-skill"
+             src="src/skills/plus.svg" width="30px" alt="adding skill">
+      </a>
+      <form class="skills__header__form" id="skills__header__form">
```
Почему форма? Я наткнулся на противоречие: инпуты должны быть в форме, но в то же время в моей форме явного сабмита не будет, я плюсик использую как сабмит выше, поэтому получилась такая недо-форма. Наверно, надо либо убрать её, либо сделать плюсик сабмит-элементом, да?

BusinessDuck replied:
лучше сделать плюсик сабмит элементом для более гибкой стилизации можно исползовать button, http://htmlbook.ru/html/button/type

avramirs replied:
Вот мне тоже так показалось, спасибо!

avramirs replied:
✅

BusinessDuck commented:
> Пока реализовано оторвано от карточки, в рамках этого дз хочу прикрутить скиллс к ней, а также залить это решение на TS, если успею, считаю, что сейчас надо упростить js, чтобы переписать потом на ts. Хочу также попросить прощения за столь позднее открытие PR'а, но надеюсь, что ещё не поздно. Страница подключена к gh-pages: https://ok-technopolis-2021.github.io/Kovalenko_Pavel/

На тайпскрипт поздно переписывать, я всем кто сдавал в первую волну проставил комменты и уже сейчас собирался выставлять финальные баллы

BusinessDuck commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+
+class Skill {
+
+    #skillName;
```
за всю историю своей работы ни разу не встречал # символ в названии атрибута, думаю ни один стандарт кода это не приветствует

avramirs replied:
Решил привнести инкапсуляции))) Идея была защиты от плохих дядь, которые получат доступ к нашим приватным переменным и начнут мутить свои дела. Но щас понял, что им труда не составляет написать this.#`<var name>`, так что смысла оно уже не имеет

avramirs replied:
✅

BusinessDuck replied:
Вопрос снят

BusinessDuck commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+    #skillName;
+    #skillRatio;
+    static #skillNameInput = document.getElementsByClassName("skills__header__form__skill-name-input")[0];
+    static #skillRatioInput = document.getElementsByClassName("skills__header__form__skill-ratio-input")[0];
```
не проще ли передать элемент снаружи? Если измениться шаблон - придется переделывать хардкод в модели скилов. Уже много раз говорил про document.querySelector, почему-то пользуются только селекторами по классу, он и выглядит не красво и предназначен для получения группы элементов

avramirs replied:
✅

BusinessDuck commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+        let skillRatio = this.#readSkillRatio();
+        if (skillName === undefined || skillRatio === undefined)
+            return undefined;
+        return new Skill(skillName, skillRatio);
```
Статичный метод класса который создает инстанс класса? Как-то сложно...

avramirs replied:
Я так на плюсах постоянно делал - сначала читаю, проверяю данные для моей структуры, а только потом создаю её с валидными параметрами. Но там был operator>>, в который надо было кидать объект, в который будут записываться данные. Наверно, тут надо сделать так же, подправлю логику

avramirs replied:
✅

BusinessDuck commented:
Посмотрел на github pages, спасибо за выкладку. 

Мне нравится, что используются формы тк это похволяет работать стандартным системным клавишам, TAB для переключения между полями ввода. Но без submit инпута, не подхватится обработка ENTER, так что он точно нужен.

Мне не нравится реализация, все выглядит очень кучно и набито в один класс. Вот как например может выглядеть архитектура такого приложения:
Класс Skill - с 2мя полями имя и уровень. Ну и еще может быть идентификатор. 
SkillList - Класс списка скилов с возможностью добавить в него скилл или удалить содержит в себе только логику работы со списком скилов, но ничего не знает про UI
SkillListController - контроллер с логикой добавления и удаления обработкой взаимодействия с юзером и логики
SkillListView - вью, для SkillList, отрисовывает класс SkillList в указаное место. 

Как то так.


Я еще дам 2 дня на правки и будет финальное оценивание.

BusinessDuck reviewed (changes requested):
все отписал в комментариях, нужны доработки

BusinessDuck commented:
> Изначально js делал в двух файлах - skill.js и skillStorage.js, однако возникла проблема с тем, что skillStorage нужен был класс skill, но заимпортить я его никак не смог - пытался и просто import "...", и @import("..."), однако ничего не помогало. В итоге разбил один класс - skillStorage - с концами, ибо два класса в файле не айс

// moduleA.js

export class ModuleA {}

// moduleB.js

export classModuleB {}

// entry point (index.js)

import { ModuleA } from './moduleA';
import { ModuleB } from './moduleB';

const moduleA = new ModuleA();
const moduleB = new ModuleB();

Вроде с импортами вот так должно стать понятнее как жить.

BusinessDuck commented on src/skills/skills.js:
```diff
@@ -0,0 +1,161 @@
+            skillName = skillName.substring(0, Skill.#skillNameInput.size);
+            alert("Skills name is very long. It will be shortened to " + skillName);
+        }
+        skillName.replaceAll("<", "&lt");
```
Создай утилиту для этого отдельную и вызови ее тут предварительн о импортировав

PaaavelZ replied:
✅

BusinessDuck reviewed (approved):
Реализация выглядит хорошо, вот для справки мой пример на ООП , всем бросаю

Сейчас в основном применяется в разработке реактивный подход. Это когда состояние модели описано отдельно, а  UI изменяется в зависимости от изменения модели. То есть у тебя есть массив данных в нашем случае это скилы, в него добавляются или удаляются скилы, и в зависимости от вида манипуляции с моделью DOM реагирует на это и перестраивается, вот ниже пример как этого можно добиться. Однако будет ряд трудностей, во первых не понятно как грамотно перерисовывать содержимое, потому что может удалиться элемент в середине, и как-то нужно вычислить, какой это элемент и удалить его точечно, также элемент может добавиться в конец или в начало (не в нашем случае а в общем). Все эти перерисовки сложно управляемы. Эту задачу решают фреймоврки как раз.

В общем и целом, реактивных подход позволяет хранить данные модели отдельно, тем самым получая к ним доступ из любых мест приложения и делая его более гибким. Для отслеживания изменений в модели чаще всего используется паттерн Observer, ссылка и описание ниже.


Вот еще несколько примеров реализации от меня

```typescript
export class Skill {
    constructor(readonly name: string, readonly value: number) {}
}

export class SkillList {
    private list: Array<Skill>;

    constructor() {
        this.list = [];
    }

    add(skill: Skill) {
        this.list.push(skill);
    }

    remove(skill: Skill) {
        this.list.splice(this.list.indexOf(skill), 1);
    }
}

export class SkillView {
    constructor(private skill: Skill) {}

    render() {
        return this.HTML(`<div class="skill">${this.skill.name}<span>${this.skill.value}</span></div>`);
    }

    private HTML(str: string) {
        // @ts-ignore
        return repalceHtmlEntities(str);
    }
}

export class SkillListView {
    private list: Array<SkillView>;

    render() {
        return this.list.map(skill => skill.render()).join('');
    }
}

export class SkillListController {
    private skillList: SkillList;
    private skillListView: SkillListView;

    constructor(private target: Element) {...}

    addSkill(...) {
        // ...
        const skill = new Skill('123', 33);
        this.skillList.add(skill);
        this.target.innerHTML = this.skillListView.render();
    }
}

// index.ts

const target = document.querySelector('.target');
const skillController = new SkillListController(target);

// В идеале вообще использовать Observer Pattern

// index2.ts
const target = document.querySelector('.target');
const model = new SkillList();
const skillController = new SkillListController(target, model);

function addSkill() {
    model.push({ name: 'asdf', value: 33 });
}

function removeSkill(idx) {
    model.removeSkill(idx);
}

// И где-то внутри контроллера скилов
// Мы подписались на изменения модели, то есть при добавлении и удалении элемента вызовется коллбек. Паттерн 
 Observer можно посмотреть тут
https://monsterlessons.com/project/lessons/observer-pattern-v-javascript
this.model.subscribe((model => {});
```

avramirs committed:
Add/Fix:
1) Shift to the limits of the skill ratio, if it is outside them;
2) Use DOMParser for readability's increasing;
3) Change addSkillBtn functionality - make it submit element;
4) Use querySelector instead array of elements by className when receiving certain input field value;
5) Simplify skills' reading.

avramirs committed:
Try to fix problem with displaying plus.svg

PaaavelZ commented:
gh-pages как-то не очень хорошо себя чувствует, мб надо больше времени на обновление. В общем, плюсик я перенёс в низ формы, локально он отображается, тут почему-то нет, но некритично, думаю. Сейчас ещё будут прикручивать к карточке, там оно всё в докс будет, думаю, всё получится.
![image](https://user-images.githubusercontent.com/76079492/146063605-24aaa4dd-414a-4a4e-8337-b2fa5c24707e.png)
![image](https://user-images.githubusercontent.com/76079492/146063701-ea61cf37-9c15-4208-9572-f5a0b7dd5701.png)

avramirs commented:
Все замечания исправлены, кроме импортов. Чтобы разбить файл через импорты, нужен галпфайл и прикрутка к основному проекту (карточке), на что у меня, к сожалению, уже нет времени из-за сессии. Это финальный вариант.

BusinessDuck reviewed (approved):
Что понравилось
1) Есть XSS защита
2) Использование приватных полей `#`
3) Сочетание ООП и функций
4) Работают горячие клавиши при добавлении

Что не понравилось
1) Статичные методы лучше было бы сделать функциями и не использовать класс, в целом показалось что не заслужено они как-то используются 
2) Стиль и верстка не очень
3) Все на отдельном экране без визитки
4) Нет import/export

---- 
Итого: 28/40
